### ISSUE TITLE
other mqtt channel

### ISSUE THREAD

**Issue by voicevon** [OttoWinter/esphomelib#319]
**Is your feature request related to a problem? Please describe.**
I have an sensor far away from my house(let's say 1km distance).There is no wifi, even no AC  power.
**Describe the solution you'd like**
Will esphomelib support LuRaWan , or GSM  connection to server?

**Additional context**
Right now, esphomelib has organized all data to an object `mqtt_message`(include topic ,payload, qos, etc.), Then it called publish() method in `mqtt_client_`  which is an AsyncMqttClient  without source code, but inside depended library. Obviously the `mqtt_client` binds AsyncTcp+Wifi. I would like to name it "Channel".

GSM module connect  to esp32 through UART (another channel),  All communication between the esp32 and GSM module  is AT-commands. 

Imaging Esphomelib `mqtt component ` has a member called `channel`. Just modify the value of `channel`  form `wifi` to `gsm`.  The esp32 will turn to  the new channel.  Because I have tested the node under `channel_wifi`, so the node should be no issue on other channel.   Isn't it awesome?

**Comment by OttoWinter** [comment]
Yes theoretically that would be possible, but it's not a focus of this project at this time - and if this happens then with the native esphomelib api (#322).

One problem with your "channel" simplification is that most protocols don't just need this abstraction, they also need the abstraction of connections. For example while MQTT theoretically can work with any channel you like, it still mandates that this "channel" have a few features:

 - Error correction
 - Closing/Opening a connection (like a TCP socket - not possible directly through UART)

So this would require another abstraction on top of "channels".

This will definitely not happen _any_ time soon (and by that I mean probably not in the next year or two) as it's just a _ton_ of work to get it to work, with honestly just a few use-cases. Esphomelib is made for homes, not for 1km distances.
